cmd/spiffe-aws-assume-role/cli/cli_test.go

```diff
@@ -0,0 +1,98 @@
+package cli
+
+import (
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/square/spiffe-aws-assume-role/internal/mocks"
+	"github.com/square/spiffe-aws-assume-role/internal/test"
+	"github.com/square/spiffe-aws-assume-role/pkg/credentials"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"testing"
+	"time"
+)
+
+var NoContext *CliContext = nil
+
+func TestParsesSessionDuration(t *testing.T) {
+	args := []string{
+		"credentials",
+		"--session-duration=PT5M",
+		// We only specify the following fields because they're required
+		"--audience=foo",
+		"--role-arn=bar",
+		"--spiffe-id=baz",
+	}
+	context, err := parse(args)
+	require.NoError(t, err)
+
+	cli := context.Model.Target.Interface().(CLI)
+	sessionDuration := cli.Credentials.SessionDuration
+	require.EqualValues(t, 5, sessionDuration.Minutes())
+}
+
+func TestDisplaysTopLevelHelp(t *testing.T) {
+	args := []string{"--help"}
+	require.NoError(t, Run(args, NoContext))
+}
+
+func TestDisplaysCredentialsHelp(t *testing.T) {
+	args := []string{"--help", "credentials"}
+	require.NoError(t, Run(args, NoContext))
+}
+
+func TestSetsCustomSessionDuration(t *testing.T) {
+	args := []string{
+		"credentials",
+		"--audience=foo",
+		"--role-arn=arn:aws:iam:123456789012:role/foo",
+		"--session-duration=PT5M",
+		"--spiffe-id=spiffe://foo",
+		"--workload-socket=tcp://127.0.0.1:8080",
+	}
+
+	jwtSource := mocks.JWTSource{}
+	defer jwtSource.AssertExpectations(t)
+	jwtSource.
+		On("FetchToken", mock.Anything).
+		Return("token", nil)
+
+	stsClient := mocks.STSAPI{}
+	defer stsClient.AssertExpectations(t)
+
+	expiration := time.Now()
+	emptyString := ""
+
+	stsCredentials := sts.Credentials{
+		AccessKeyId:     &emptyString,
+		Expiration:      &expiration,
+		SecretAccessKey: &emptyString,
+		SessionToken:    &emptyString,
+	}
+
+	output := sts.AssumeRoleWithWebIdentityOutput{
+		Credentials: &stsCredentials,
+	}
+
+	captor := test.Captor{}
+
+	stsClient.
+		On("AssumeRoleWithWebIdentityWithContext", mock.Anything, mock.Anything).
+		Run(captor.Capture()).
+		Return(&output, nil)
+
+	context := CliContext{
+		JWTSourceProvider: credentials.StaticJWTSourceProvider(&jwtSource),
+		STSProvider:       credentials.StaticSTSProvider(&stsClient),
+	}
+
+	failOnError(t, Run(args, &context))
+
+	input := captor.Args.Get(1).(*sts.AssumeRoleWithWebIdentityInput)
+	require.EqualValues(t, 5, *input.DurationSeconds)
+}
```

Comment:
I wrote up a little Captor helper class which allows us to capture and inspect arguments in Mockito-esque style. Maybe there's a built-in way to accomplish something similar but if so it eluded me.

Reply:
I think testify expects you to be explicit about the arguments in 
```
stsClient.
	On("AssumeRoleWithWebIdentityWithContext", mock.Anything, mock.Anything)
``` 
if you wanted to verify that the function was called correctly. Rather than capturing the arguments, you would explicitly define the `AssumeRoleWithWebIdentityInput` you're expecting instead of using `mock.Anything`

Reply:
My only complaint there is that, by having to fabricate the entire structure and have the whole thing match, this test becomes more brittle in the face of unrelated future changes. By capturing the structure and only inspecting the bits I care about it's more future-proof.